generate.py

```diff
@@ -50,19 +61,22 @@ def generate(
         idx_cond = idx_cond if t <= max_seq_length else idx_cond[-max_seq_length:]
 
         # forward
-        logits = model(idx_cond.view(1, -1))
+        logits, cache_kvs = model(input_idx.view(1, -1), input_pos, cache_kvs)
         logits = logits[0, -1] / temperature
 
         # optionally crop the logits to only the top k options
         if top_k is not None:
             v, _ = torch.topk(logits, min(top_k, logits.size(-1)))
-            logits[logits < v[[-1]]] = -float("Inf")
+            logits = torch.where(logits < v[[-1]], -float("Inf"), logits)
 
         probs = torch.nn.functional.softmax(logits, dim=-1)
         idx_next = torch.multinomial(probs, num_samples=1)
```

Comment:
index_copy requires that the dtype matches, so we need

```suggestion
        idx_next = torch.multinomial(probs, num_samples=1).to(dtype=idx.dtype)
```